### PR TITLE
Add support for nerdctl rmi.

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan <smahajan@roblox.com>

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
```
root@vagrant:~/go/src/github.com/nerdctl# nerdctl rmi traefik redis
Untagged: traefik:latest
Deleted: sha256:3e207b409db364b595ba862cdc12be96dcdad8e36c59a03b7b3b61c946a5741a
Deleted: sha256:4dca0fb1912d95147c68fab2028b462575694acfcbce3cc1580d791619f762ec
Deleted: sha256:ec42ef6229b5f723d9f61d8383ef8e001e0d656392ef8bda323158172a1c1127
Deleted: sha256:04237b2012552f2252018027944b634ca2f5cf11c1020722fd8d29997920420e
Untagged: redis:latest
Deleted: sha256:f5600c6330da7bb112776ba067a32a9c20842d6ecc8ee3289f1a713b644092f8
Deleted: sha256:c492b4fe510f674e04af29c5a8be4d0104a53e82e68ae98a3f9a0dd385263b7c
Deleted: sha256:60f4b166131fd0c3b925a04e4976fbc43525e289e5edab72ffc4b7196085905f
Deleted: sha256:4bc5e04ae88961aa002128a1873888ad4adf5ce11712d601fec6ad5f280c547d
Deleted: sha256:06d5c35d63a8e4066ca996005b118a6051d2323c4af64c0fc2be014b4da7740e
Deleted: sha256:d3fc9f08141a1377e718b8a29fd94a8c46793b88efdbed4ad02cbe3a3d91de10
root@vagrant:~/go/src/github.com/nerdctl# nerdctl rmi traefik redis
Error: No such image: traefik
Error: No such image: redis
root@vagrant:~/go/src/github.com/nerdctl# echo $?
1
root@vagrant:~/go/src/github.com/nerdctl#
```

/cc @AkihiroSuda

File: rmi.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+				return nil, err
+			}
+			break
+		}
+	}
+	return digests, nil
+}
+
+// Get fully qualified image name (FQIN) for the given image.
+func getFQIN(img string) string {
```
Comment (AkihiroSuda) [inline rmi.go:145]: Please use this: https://github.com/AkihiroSuda/nerdctl/blob/ff45df2f151dd25ff9863ce6d154013141555d5b/pkg/imgutil/imgutil.go#L45

File: rmi.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		fmt.Printf("Deleted: %s\n", digest)
+	}
+}
+
+// Get SHA digests for the given image.
+func getImageDigests(ctx context.Context, cs content.Store, imgFQIN string, imageList []images.Image) ([]digest.Digest, error) {
+	var digests []digest.Digest
+	var err error
+	for _, image := range imageList {
+		if imgFQIN == image.Name {
```
Comment (AkihiroSuda) [inline rmi.go:133]: I guess we want to support short ID as in `nerdctl rm`, but can be a separate PR

File: rmi.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	return nil
+}
+
+// Print digests after image removal.
+// This will keep the stdout in sync with docker rmi output.
+func printDigests(imgFQIN string, digests []digest.Digest) {
+	if strings.Contains(imgFQIN, "docker.io/library") {
+		imgFQIN = imgFQIN[18:]
+	}
+	fmt.Printf("Untagged: %s\n", imgFQIN)
```
Comment (AkihiroSuda) [inline rmi.go:122]: `fmt.Fprintf(clicontext.App.Writer, ...)`

File: rmi.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		imgFQIN := getFQIN(img)
+
+		digests, err := getImageDigests(ctx, cs, imgFQIN, imageList)
+		if err != nil {
+			return errors.Errorf("Error in getting image digests: %v", err)
+		}
+
+		if err := imageStore.Delete(ctx, imgFQIN, opts...); err != nil {
+			if errdefs.IsNotFound(err) {
+				fmt.Printf("Error: No such image: %s\n", img)
```
Comment (AkihiroSuda) [inline rmi.go:100]: logrus.WithError(err).Errorf(...)

Comment (shishir-a412ed) [reply rmi.go:100]: So I made this change, however when the images are missing the error looks like this:

```
root@vagrant:~/go/src/github.com/nerdctl# nerdctl rmi redis traefik:v2.3.2
ERRO[0000] Error: No such image: redis                   error="image \"docker.io/library/redis:latest\": not found"
ERRO[0000] Error: No such image: traefik:v2.3.2          error="image \"docker.io/library/traefik:v2.3.2\": not found"
```

My reason for using `fmt.Printf()` was to keep the output the same as `docker rmi`. This is how it looks like in `docker rmi`

```
 SM-smahajan628:~ smahajan$ docker rmi traefik redis
Error: No such image: traefik
Error: No such image: redis
SM-smahajan628:~ smahajan$ echo $?
1
```

Comment (AkihiroSuda) [reply rmi.go:100]: SGTM, but `fmt.Fprintf(clicontext.App.Writer, ...)` would be preferable than plain `fmt.Printf`

File: rmi.go (hunk 1 above)
Comment (shishir-a412ed) [reply rmi.go]: Fixed.

File: rmi.go (hunk 3 above)
Comment (shishir-a412ed) [reply rmi.go]: Fixed.

File: rmi.go (hunk 4 above)
Comment (shishir-a412ed) [reply rmi.go]: Fixed.

File: rmi.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	if clicontext.NArg() == 1 {
+		if imageNotFoundError {
+			os.Exit(1)
+		}
+		os.Exit(0)
+	}
+
+	for _, img := range clicontext.Args().Tail() {
```
Comment (AkihiroSuda) [inline rmi.go:96]: The loop seems redundant.

Can we write like this:
```go
for i, img := range clicontext.Args() {
  doStuff()
  if i == 0 && imageNotFoundError != nil {
    return imageNotFoundError
  }
}
```

Comment (shishir-a412ed) [reply rmi.go:96]: Are you able to range over `clicontext.Args()`?

I added this to try it out

```
root@vagrant:~/go/src/github.com/nerdctl# git diff
diff --git a/rmi.go b/rmi.go
index 229ee2a..2b80e3c 100644
--- a/rmi.go
+++ b/rmi.go
@@ -64,6 +64,12 @@ func rmiAction(clicontext *cli.Context) error {
        var opts []images.DeleteOpt
        var imageNotFoundError bool

+       for i, img := range clicontext.Args() {
+               fmt.Println(i+" "+img)
+       }
+
+       os.Exit(0)
+
        if clicontext.NArg() >= 1 {
                img := clicontext.Args().First()
                named, err := refdocker.ParseDockerRef(img)
```

and getting the following error on `make`

```
./rmi.go:67:16: cannot range over clicontext.Args() (type "github.com/urfave/cli/v2".Args)
Makefile:39: recipe for target 'nerdctl' failed
make: *** [nerdctl] Error 2
```

Comment (shishir-a412ed) [reply rmi.go:96]: https://godoc.org/gopkg.in/urfave/cli.v2#Args

Comment (shishir-a412ed) [reply rmi.go:96]: Maybe `clicontext.Args().Get(i)`? Trying it now ....

Comment (AkihiroSuda) [reply rmi.go:96]: Args().Slice() returns []string

Comment (shishir-a412ed) [reply rmi.go]: Fixed.

**Review by AkihiroSuda — APPROVED** [review]
Thanks
